### ISSUE TITLE
Feat: "Create project" from blueprint (or template)

### ISSUE THREAD

**Issue by michaelbrewer** [coder/coder#12510]
If developers are to log into Coder from day 1, it would help that they can create new projects from which they can open in a coder workspace with assurances that it would work and be compliant with the companies best practices.

CodeCatalyst has is a blueprint catalog and you can make custom blueprints (https://codecatalyst.aws/explore/blueprints) and CodeSpaces has something similar with templates (https://github.com/codespaces/templates). 

Below screen shot is from CodeSpaces.

<img width="1727" alt="Screenshot 2024-03-11 at 1 21 09 AM" src="https://github.com/coder/coder/assets/5442469/25bb1271-a2a6-415e-8ef5-e88278c8c0ba">

## Proposal

Add support for creating new projects from blueprints inside of Coder to further simplify the developer experience.

---

## Alternatives

Creating projects from a template before going to Coder (like from Backstage, Gitlab or GIhub). However then you are moving from one tool to another.

If coder can be fully integrated into an IDP like backstage (https://backstage.io/docs/features/software-templates/) then you make get a similar feel

**Comment by matifali** [comment]
How are blueprints different from Coder [templates](https://coder.com/docs/v2/latest/templates/creating)?

**Comment by michaelbrewer** [comment]
> How are blueprints different from Coder [templates](https://coder.com/docs/v2/latest/templates/creating)?

Coder templates allows you to create a workspace. 

Blueprints here is to create new project (before it even exists in an repository).

So the workflow could be:

- Create project -> Select Blueprint -> Open in workspace

Now the generated project would need to be either stashed somewhere before being loaded into the workspace, or it would actually initialize against a selected external authentication.

**Comment by matifali** [comment]
I think we can relate this to #8462 or our upcoming #10656

**Comment by michaelbrewer** [comment]
I can see how #8462 especially considering the wireframes (https://github.com/coder/coder/issues/8462#issuecomment-1816690052)

Backstage integration would allow you to create new projects from their blueprints, but then you would need to be running both Backstage and Coder to leverage this feature. CodeCatalyst takes an all in one approach, supporting blueprints that create new projects and then you can open projects in a Dev Environment, all without leaving that one tool (or logging into another tool).

Let me think before closing this issue :)

**Comment by matifali** [comment]
Also we do have [Start Templates](https://github.com/coder/coder/tree/main/examples/templates) that let you create a template in a fresh installation (configuration required). Isn't this analogous to Blueprints?  They are defined here:

https://github.com/coder/coder/blob/5296611a3fa3a3524181182444fffd7c51030341/examples/examples.go#L25-L38

![image](https://github.com/coder/coder/assets/10648092/40640a61-5b05-4680-9db8-b9e08aad5481)

**Comment by michaelbrewer** [comment]
@matifali - can interested in the "create new project" flow vs the "create a workspace".

- "create new project" - Is before you have a repo and want to build a new application, like a springboot application. Here patterns would be used and stored as a git repo or using something like "Spring initialzer"

- "create new workspace" - is from a "Template" (which is confusingly names in the context of project templates) and creates the compute environment in which you can open a project, or create a new one (but now it is after that fact, did i select the right compute image to start this new project?)

- "create coder template" - is the host terraform template for provisioning workspaces.

**Comment by michaelbrewer** [comment]
Interesting that `Devfile` spec supports this : https://registry.devfile.io/viewer/devfiles/community/dotnet60

```yaml
starterProjects:
  - name: dotnet60-example
    git:
      checkoutFrom:
        remote: origin
        revision: dotnet-6.0
      remotes:
        origin: https://github.com/redhat-developer/s2i-dotnetcore-ex
    subDir: app
```

**Comment by brunnels** [comment]
> Interesting that `Devfile` spec supports this : https://registry.devfile.io/viewer/devfiles/community/dotnet60
> 
> ```yaml
> starterProjects:
>   - name: dotnet60-example
>     git:
>       checkoutFrom:
>         remote: origin
>         revision: dotnet-6.0
>       remotes:
>         origin: https://github.com/redhat-developer/s2i-dotnetcore-ex
>     subDir: app
> ```

Yes, I would be very interested if coder could leverage devfiles as they are well defined and there's also an operator to deploy them.   https://github.com/devfile/devworkspace-operator
